includes/rest/class-outbox-controller.php

```diff
@@ -0,0 +1,284 @@
+			'partOf'       => \get_rest_url( null, \sprintf( 'actors/%d/outbox', $user_id ) ),
+			'totalItems'   => $outbox_query->found_posts,
+			'orderedItems' => array(),
+		);
+
+		foreach ( $query_result as $outbox_item ) {
+			$response['orderedItems'][] = $this->prepare_item_for_response( $outbox_item, $request );
+		}
+
+		$post_types = \get_option( 'activitypub_support_post_types', array( 'post' ) );
+		if ( $user_id > 0 ) {
+			$count_posts            = \count_user_posts( $user_id, $post_types, true );
+			$response['totalItems'] = \intval( $count_posts );
+		} else {
+			foreach ( $post_types as $post_type ) {
+				$count_posts             = \wp_count_posts( $post_type );
+				$response['totalItems'] += \intval( $count_posts->publish );
+			}
+		}
+
+		$max_pages         = \ceil( $response['totalItems'] / $request->get_param( 'per_page' ) );
+		$response['first'] = \add_query_arg( 'page', 1, $response['partOf'] );
+		$response['last']  = \add_query_arg( 'page', $max_pages, $response['partOf'] );
+
+		if ( $max_pages > $page ) {
+			$response['next'] = \add_query_arg( 'page', $page + 1, $response['partOf'] );
+		}
+
+		if ( $page > 1 ) {
+			$response['prev'] = \add_query_arg( 'page', $page - 1, $response['partOf'] );
+		}
+
+		if ( $page > $max_pages && $response['totalItems'] > 0 ) {
+			return new \WP_Error(
+				'rest_post_invalid_page_number',
+				'The page number requested is larger than the number of pages available.',
+				array( 'status' => 400 )
+			);
+		}
+
+		/**
+		 * Filter the ActivityPub outbox array.
+		 *
+		 * @param array $response The ActivityPub outbox array.
+		 */
+		$response = \apply_filters( 'activitypub_rest_outbox_array', $response );
+
+		/**
+		 * Action triggered after the ActivityPub profile has been created and sent to the client.
+		 */
+		\do_action( 'activitypub_outbox_post' );
+
+		$response = \rest_ensure_response( $response );
+		$response->header( 'Content-Type', 'application/activity+json; charset=' . \get_option( 'blog_charset' ) );
+
+		return $response;
+	}
+
+	/**
+	 * Validates the user_id parameter.
+	 *
+	 * @param mixed $user_id The user_id parameter.
+	 * @return bool|\WP_Error True if the user_id is valid, WP_Error otherwise.
+	 */
+	public function validate_user_id( $user_id ) {
+		$user = Actors::get_by_various( $user_id );
+		if ( \is_wp_error( $user ) ) {
+			return $user;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Prepares the item for the REST response.
+	 *
+	 * @param mixed            $item    WordPress representation of the item.
+	 * @param \WP_REST_Request $request Request object.
+	 * @return array Response object on success, or WP_Error object on failure.
+	 */
+	public function prepare_item_for_response( $item, $request ) {
+		$transformer = Factory::get_transformer( $item->post_content );
+
+		$type  = 'Object';
+		$terms = wp_get_object_terms( $item->ID, 'ap_activity_type' );
+		if ( isset( $terms[0]->name ) ) {
+			$type = ucfirst( $terms[0]->name );
+		}
```

Comment:
These lines don't feel very reassuring. Is there a way to store the activity type with the outbox item itself somehow?

Reply:
I planned the `ap_activity_type` taxonomy for that: https://github.com/Automattic/wordpress-activitypub/pull/593/files#diff-480b2542905cc428144536831a25aa56f1c4307d2667a40f719353b6ed91bf8dR574